Added definition `function_module(f)`

mauro3 opened the pull request:
This fixes #14636, as promised.

Also added equivalent function `datatype_module(dt)`, moved `function_name` to a better place further down in the file, and converted to in-line doc-strings.

Problem, when generating the documentation I get the warning:

```
WARNING: missing docs for signature:

    function_module(f::Function, types) -> Module
```

mauro3 committed:
Added definition function_module(f)

Also added equivalent function datatype_module(dt)

MichaelHatherly commented on base/reflection.jl:
```diff
@@ -450,6 +457,19 @@ function functionloc(f)
+function_module(f) = datatype_module(f)
+
+"""
+  function_module(f::Function, types) -> Module
```
Needs a four space indent for markdown to parse it as a code block. (That's where the `genstdlib.jl` warning is coming from I think.)

martinholters commented on base/reflection.jl:
```diff
@@ -74,6 +74,13 @@ end
+
+Determine the module containing the definition of a `DataType`.
+"""
+datatype_module(dt) = typeof(dt).name.module
```
My interpretation of the docstring is that `dt` is the type in question and not an instance of it, e.g. `datatype_module(Range)` should yield `Base`. But then the `typeof` is too much. (`typeof(Range).name.module` would be `DataType.name.module`, namely `Core`).

martinholters commented:
Tests?

mauro3 commented:
Thanks for the feedback! I fixed the doc-string (which got rid of the warning), changed the definition of `datatype_module` and added tests.

tkelman commented:
if these are present in the rst docs, can you run `make docs` and commit the change if there is any?

tkelman commented on base/reflection.jl:
```diff
@@ -450,6 +457,19 @@ function functionloc(f)
 end
 
+"""
+    function_module(f) -> Module
```
if the signature below is in the rst docs, then this one probably should be too

mauro3 replied:
You mean it should be in  https://github.com/JuliaLang/julia/blob/master/doc/stdlib/base.rst, right?  That file is auto-generated, but running `make julia-genstdlib` doesn't do anything. What do I need to do?

tkelman replied:
the file is not autogenerated. only the docstring parts of it that say they are. read contributing.md, there's a section about what you need to do. add the signature to the rst then genstdlib will populate the docstring content

mauro3 commented:
Thanks Tony, fixed. I also changed the signature to `function_module(f::Function)` to be in-line with `function_module(f::Function,types)`.

MichaelHatherly commented:
Should `datatype_module` be added to the rst docs as well or is it already in there somewhere?

tkelman commented:
which of these are exported?

mauro3 commented:
None are exported.  `datatype_module` is a new function.  It seems logical to include it.  I can add it to the rst too.  But not, not all functions in reflections.jl are in the rst.

tkelman commented:
hm then maybe the docstrings should say `Base.` ? I think we should have a Reflection module at some point.

mauro3 commented:
Unless I hear differently, I will add a `Base.` in the docs to all unexported methods in reflection.jl.  I will also add `datatype_module` to the rst docs. (after the weekend, that is)

mauro3 commented:
Added the `Base.` to the docstrings and rst.  Added docs for `datatype_module` to the rst. `function_name` also was missing a `Base.` in its doc-string which I added too.

tkelman commented:
needs rebase, but I think this was okay. not exported so don't think this is a big enough deal to need to wait until after branching

mauro3 commented:
Rebased.  There is no hurry on my side with getting this merged.  Although, if deemed fine to merge now, all the better.
